crates/ruff_linter/src/suppression.rs

```diff
@@ -159,82 +166,140 @@ impl Suppressions {
+                        context
+                            .report_diagnostic(
+                                InvalidRuleCode {
+                                    rule_code: suppression.code.to_string(),
+                                    kind: InvalidRuleCodeKind::Suppression,
+                                },
+                                range,
+                            )
+                            .set_fix(Fix::safe_edit(edit));
                     }
-                } else {
-                    UnusedCodes {
-                        disabled: vec![suppression.code.to_string()],
-                        ..Default::default()
+                }
+            } else if !suppression.used.get() {
+                // UnusedNOQA
+                if context.is_rule_enabled(Rule::UnusedNOQA) {
+                    let Ok(rule) = Rule::from_code(
+                        get_redirect_target(&suppression.code).unwrap_or(&suppression.code),
+                    ) else {
+                        continue; // should trigger InvalidRuleCode above
+                    };
+                    for comment in &suppression.comments {
+                        let (range, edit) =
+                            Suppressions::delete_code_or_comment(locator, suppression, comment);
+
+                        let codes = if context.is_rule_enabled(rule) {
+                            UnusedCodes {
+                                unmatched: vec![suppression.code.to_string()],
+                                ..Default::default()
+                            }
+                        } else {
+                            UnusedCodes {
+                                disabled: vec![suppression.code.to_string()],
+                                ..Default::default()
+                            }
+                        };
+
+                        context
+                            .report_diagnostic(
+                                UnusedNOQA {
+                                    codes: Some(codes),
+                                    kind: UnusedNOQAKind::Suppression,
+                                },
+                                range,
+                            )
+                            .set_fix(Fix::safe_edit(edit));
                     }
-                };
+                }
+            } else if suppression.comments.len() == 1 {
+                // UnmatchedSuppressionComment
+                let range = suppression.comments[0].range;
+                if unmatched_ranges.insert(range) {
+                    context.report_diagnostic_if_enabled(UnmatchedSuppressionComment {}, range);
+                }
+            }
+        }
 
-                let mut diagnostic = context.report_diagnostic(
+        for error in &self.errors {
+            // treat comments with no codes as unused suppression
+            if error.kind == ParseErrorKind::MissingCodes {
```

Comment:
I wonder if this is confusing for users that expect `ruff: disable` to work without rule codes, and having it report a invalid syntax error would be more helpful in that case (we could even special case it with a help text, saying they should add a rule code)